op-supervisor/supervisor/backend/processors/chain_processor.go

```diff
@@ -151,34 +155,116 @@ func (s *ChainProcessor) work() {
+	next := s.nextNum()
+	last := s.lastHead.Load()
+	// next is already beyond the end, nothing to do
+	if next > last {
+		return 0, nil
+	}
+	nums := make([]uint64, 0)
+	for i := next; i <= last; i++ {
+		nums = append(nums, i)
+		// only collect as many blocks as we can fetch in parallel
+		if len(nums) >= s.maxFetcherThreads {
+			break
+		}
+	}
+
+	s.log.Debug("Fetching blocks", "chain", s.chain.String(), "next", next, "last", last, "count", len(nums))
+
+	// make a structure to receive parallel results
+	type keyedResult struct {
+		num      uint64
+		blockRef *eth.BlockRef
+		receipts gethtypes.Receipts
+		err      error
+	}
+	parallelResults := make(chan keyedResult, len(nums))
+
+	// each thread will fetch a block and its receipts and send the result to the channel
+	fetch := func(wg *sync.WaitGroup, num uint64) {
+		defer wg.Done()
+		// ensure we emit the result at the end
+		result := keyedResult{num, nil, nil, nil}
+		defer func() { parallelResults <- result }()
+
+		// fetch the block ref
+		ctx, cancel := context.WithTimeout(s.ctx, time.Second*10)
+		nextL1, err := s.client.L1BlockRefByNumber(ctx, num)
+		cancel()
+		if err != nil {
+			result.err = err
+			return
+		}
+		next := eth.BlockRef{
+			Hash:       nextL1.Hash,
+			ParentHash: nextL1.ParentHash,
+			Number:     nextL1.Number,
+			Time:       nextL1.Time,
+		}
+		result.blockRef = &next
+
+		// fetch receipts
+		ctx, cancel = context.WithTimeout(s.ctx, time.Second*10)
+		_, receipts, err := s.client.FetchReceipts(ctx, next.Hash)
+		cancel()
+		if err != nil {
+			result.err = err
+			return
+		}
+		result.receipts = receipts
 	}
 
-	ctx, cancel := context.WithTimeout(s.ctx, time.Second*10)
-	nextL1, err := s.client.L1BlockRefByNumber(ctx, nextNum)
-	next := eth.BlockRef{
-		Hash:       nextL1.Hash,
-		ParentHash: nextL1.ParentHash,
-		Number:     nextL1.Number,
-		Time:       nextL1.Time,
+	// kick off the fetches and wait for them to complete
+	var wg sync.WaitGroup
+	for _, num := range nums {
+		wg.Add(1)
+		go fetch(&wg, num)
 	}
-	cancel()
-	if err != nil {
-		return fmt.Errorf("failed to fetch next block: %w", err)
+	wg.Wait()
+
+	// collect and sort the results
+	results := make([]keyedResult, len(nums))
+	for i := range nums {
+		result := <-parallelResults
+		results[i] = result
 	}
+	slices.SortFunc(results, func(a, b keyedResult) int {
+		if a.num < b.num {
+			return -1
+		}
+		if a.num > b.num {
+			return 1
+		}
+		return 0
+	})
 
-	// Try and fetch the receipts
-	ctx, cancel = context.WithTimeout(s.ctx, time.Second*10)
-	_, receipts, err := s.client.FetchReceipts(ctx, next.Hash)
-	cancel()
-	if err != nil {
-		return fmt.Errorf("failed to fetch receipts of block: %w", err)
+	// process the results in order and return the first error encountered,
+	// and the number of blocks processed successfully by this call
+	for i := range results {
+		if results[i].err != nil {
+			fmt.Println("hit an error at index", i, "err", results[i].err)
```

Comment:
looks like we want to remove this print statement

Reply:
good catch, thanks!